gcp-pubsub/src/test/groovy/io/micronaut/gcp/pubsub/bind/PubSubPushSpec.groovy

```diff
@@ -0,0 +1,271 @@
+        PubsubMessage receivedMessage = consumer.msg as PubsubMessage
+
+        then:
+        receivedMessage
+        receivedMessage.messageId == "1"
+        receivedMessage.publishTime.seconds == 1614366835L
+        receivedMessage.publishTime.nanos == 749000000L
+        receivedMessage.data
+
+        when:
+        Book receivedBook = JsonMapper.createDefault().readValue(receivedMessage.data.toByteArray(), Book.class)
+
+        then:
+        receivedBook
+        receivedBook.title == book.title
+        receivedBook.author == book.author
+    }
+
+    void "a push message with POJO data can be received as a raw byte array"() {
+        given:
+        Book book = new Book()
+        book.title = "Fight Club"
+        book.author = "Chuck Palahniuk"
+        String data = JsonMapper.createDefault().writeValueAsString(book)
+        String encodedData = Base64.getEncoder().encodeToString(data.getBytes())
+        PushRequest request = new PushRequest("projects/test-project/subscriptions/raw", new PushRequest.PushMessage(null, encodedData, "1", "2021-02-26T19:13:55.749Z"))
+
+        when:
+        HttpResponse response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request))
+
+        then:
+        response.status() == HttpStatus.OK
+        consumer.msg
+
+        when:
+        byte[] received = consumer.msg as byte[]
+        Book receivedBook = JsonMapper.createDefault().readValue(received, Book.class)
+
+        then:
+        receivedBook
+        receivedBook.title == book.title
+        receivedBook.author == book.author
+    }
+
+    @Property(name = "gcp.pubsub.push.path", value = "/custom-push-path")
+    void "a custom path may be used for push messages"() {
+        given:
+        String data = JsonMapper.createDefault().writeValueAsString(Map.of("key", "success"))
+        String encodedData = Base64.getEncoder().encodeToString(data.getBytes())
+        PushRequest request = new PushRequest("projects/test-project/subscriptions/foo", new PushRequest.PushMessage(Map.of("foo", "bar"), encodedData, "1", "2021-02-26T19:13:55.749Z"))
+
+        when:
+        HttpResponse response = pushClient.toBlocking().exchange(HttpRequest.POST("/custom-push-path", request))
+
+        then:
+        response.status() == HttpStatus.OK
+        consumer.msg
+
+        when:
+        Map<String, String> received = consumer.msg as Map<String, String>
+
+        then:
+        received.get("key") == "success"
+    }
+
+    @Property(name = "gcp.pubsub.push.enabled", value = StringUtils.FALSE)
+    void "push messaging can be disabled"() {
+        given:
+        String data = JsonMapper.createDefault().writeValueAsString(Map.of("key", "success"))
+        String encodedData = Base64.getEncoder().encodeToString(data.getBytes())
+        PushRequest request = new PushRequest("projects/test-project/subscriptions/foo", new PushRequest.PushMessage(Map.of("foo", "bar"), encodedData, "1", "2021-02-26T19:13:55.749Z"))
+
```

Comment:
If we `@Inject` a `BeanContext` into this, and fix the controller `@Requires` to use the right configuration class (see above), we can test the controller is not available when disabled with

```suggestion
        expect:
        !beanContext.containsBean(PushController)

```

Reply:
(this fails if you do not fix the controller `Requires`, so I think it's worth adding)